File: app/src/test/java/com/egecius/rxjava2_demo_2/rx/subjects/AsyncSubjectExamplesTest.java
package com.egecius.rxjava2_demo_2.rx.subjects;

import static com.egecius.rxjava2_demo_2.rx.subjects.AsyncSubjectExamples.ASYNC_EXCEPTION;

import static java.util.Arrays.asList;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.List;

import io.reactivex.observers.TestObserver;
import io.reactivex.subjects.AsyncSubject;

@RunWith(MockitoJUnitRunner.class)
public class AsyncSubjectExamplesTest {

    private AsyncSubjectExamples mSut;

    @Before
    public void setUp() {
        mSut = new AsyncSubjectExamples();
    }

    @Test
    public void emitsOnlyLastEventBeforeTermination() {

        List<Integer> list = asList(1, 2, 3);
        AsyncSubject<Integer> subject = mSut.emitEventsWithOnComplete(list);

        TestObserver<Integer> testObserver = subject.test();

        testObserver
                .assertComplete()
                .assertNoErrors()
                .assertValue(3);
    }

    @Test
    public void emitsNoEventsWhenTerminatesWithError() {

        List<Integer> list = asList(1, 2, 3);
        AsyncSubject<Integer> subject = mSut.emitEventsWithError(list);

        TestObserver<Integer> testObserver = subject.test();

        testObserver
                .assertNoValues()
                .assertError(ASYNC_EXCEPTION);
    }

    @Test
    public void emitsNoValuesWithoutTermination() {

        List<Integer> list = asList(1, 2, 3);
        AsyncSubject<Integer> subject = mSut.emitEventsWithoutTermination(list);

        TestObserver<Integer> testObserver = subject.test();

        testObserver
                .assertNoValues()
                .assertNotTerminated();
    }

}

File: app/src/test/java/com/egecius/rxjava2_demo_2/rx/subjects/AsyncSubjectExamplesTest.kt
package com.egecius.rxjava2_demo_2.rx.subjects

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner
import java.util.Arrays.asList

@RunWith(MockitoJUnitRunner::class)
class AsyncSubjectExamplesTest {

    private var mSut: AsyncSubjectExamples? = null

    @Before
    fun setUp() {
        mSut = AsyncSubjectExamples()
    }

    @Test
    fun emitsOnlyLastEventBeforeTermination() {

        val list = asList(1, 2, 3)
        val subject = mSut!!.emitEventsWithOnComplete(list)

        val testObserver = subject.test()

        testObserver
                .assertComplete()
                .assertValue(3)
    }

    @Test
    fun emitsNoEventsWhenTerminatesWithError() {

        val list = asList(1, 2, 3)
        val subject = mSut!!.emitEventsWithError(list)

        val testObserver = subject.test()

        testObserver
                .assertNoValues()
                .assertError(AsyncSubjectExamples.ASYNC_EXCEPTION)
    }

    @Test
    fun emitsNoValuesWithoutTermination() {

        val list = asList(1, 2, 3)
        val subject = mSut!!.emitEventsWithoutTermination(list)

        val testObserver = subject.test()

        testObserver
                .assertNoValues()
                .assertNotTerminated()
    }

}